### PR TITLE
[SPARK-13830] prefer block manager than direct result for large result

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

The current RPC can't handle large blocks very well, it's very slow to fetch 100M block (about 1 minute). Once switch to block manager to fetch that, it took about 10 seconds (still could be improved).
## How was this patch tested?

existing unit tests.

### CODE REVIEW

**Comment by davies** [comment]
cc @zsxwing @andrewor14

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 1]
```diff
@@ -97,9 +97,9 @@ private[spark] class Executor(
   // Set the classloader for serializer
   env.serializer.setDefaultClassLoader(replClassLoader)
 
-  // Max RPC message size. If task result is bigger than this, we use the block manager
+  // Max size of direct result. If task result is bigger than this, we use the block manager
   // to send the result back.
-  private val maxRpcMessageSize = RpcUtils.maxMessageSizeBytes(conf)
+  private val maxDirectResultSize = Math.min(1L << 20, RpcUtils.maxMessageSizeBytes(conf))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: we should have another config for this instead. Also 1MB is very small, maybe 16 or 32MB?

**Comment by andrewor14** [comment]
LGTM

**Comment by zsxwing** [comment]
LGTM

**Comment by davies** [comment]
Merged into master, thanks!

**Pull request closed without merging**